listenbrainz/webserver/static/js/src/utils.tsx

```diff
@@ -102,4 +102,67 @@ const getPlayButton = (listen: any, onClickFunction: () => void) => {
   );
 };
 
-export { searchForSpotifyTrack, getArtistLink, getTrackLink, getPlayButton };
+const formatWSMessageToListen = (wsMsg: any): Listen | null => {
+  const json = wsMsg;
+  try {
+    // The websocket message received may not contain the expected track_metadata and listened_at fields.
+    // Therefore, we look for their alias as well.
+    if (!("track_metadata" in json)) {
+      if ("data" in json) {
+        json.track_metadata = json.data;
+        delete json.data;
+      } else {
+        // eslint-disable-next-line no-console
+        console.debug(
+          "Could not find track_metadata and data in following json: ",
+          json
+        );
+        return null;
+      }
+    }
+    if (!("listened_at" in json)) {
+      if ("timestamp" in json) {
+        json.listened_at = json.timestamp;
+        delete json.timestamp;
+      } else {
+        // eslint-disable-next-line no-console
+        console.debug(
+          "Could not find listened_at and timestamp in following json: ",
+          json
+        );
+        return null;
+      }
+    }
+    // The websocket message received contains the recording_msid as a top level key.
+    // Therefore, we need to shift it json.track_metadata.additional_info.
+    if (!_.has(json, "track_metadata.additional_info.recording_msid")) {
+      if ("recording_msid" in json) {
+        json.track_metadata.additional_info.recording_msid =
```

Comment:
This line also has the potential to break. I thought I commented on it with the last review but I must have lost my comment.

Javascript won't like it if the json object doesn't have the structure we expect.
Say for example there's no `additional_info`,  this line will throw an error trying to set a chid property of `undefined`.

However like above there's an elegant solution with [lodash's merge method](http://lodash.com/docs/#merge):

```
merge(json, {
  track_metadata: { additional_info: { recording_msid: json.recording_msid } },
});
```
That should ensure the structure.

Reply:
This came to my mind but I missed it.Thanks